dot/parachain/statement-distribution/active_leaves_update.go

```diff
@@ -0,0 +1,310 @@
+			delete(s.state.unusedTopologies, sessionIdx)
+			perSessionState.supplyTopology(&top.Topology, top.LocalIndex)
+		}
+
+		s.state.perSession[sessionIdx] = perSessionState
+	}
+
+	perSession := s.state.perSession[sessionIdx]
+	if perSession == nil {
+		panic("either existed or just inserted; qed.")
+	}
+
+	if len(disableValidatorsSet) > 0 {
+		disabled := make([]parachaintypes.ValidatorIndex, len(disableValidatorsSet))
+		for d := range disableValidatorsSet {
+			disabled = append(disabled, d)
+		}
+
+		logger.Debugf("disabled validators detected: %v, "+
+			"session index=%v, relay parent=%s", disabled, sessionIdx, rp.String())
+	}
+
+	validatorGroups, err := rt.ParachainHostValidatorGroups()
+	if err != nil {
+		return fmt.Errorf("querying validator groups: %w", err)
+	}
+
+	claimQueue, err := rt.ParachainHostClaimQueue()
+	if err != nil {
+		return fmt.Errorf("querying host claim queue")
+	}
+
+	groupsPerPara, assignmentsPerGroup := determineGroupAssignment(
+		len(perSession.groups.all()),
+		&validatorGroups.GroupRotationInfo,
+		&claimQueue,
+	)
+
+	var localValidator *localValidatorState
+	if perSession.localValidator != nil {
+		localValidator = findActiveValidatorState(
+			*perSession.localValidator,
+			perSession.groups,
+			assignmentsPerGroup,
+		)
+	} else {
+		localValidator = &localValidatorState{
+			gridTracker: newGridTracker(),
+			active:      nil,
+		}
+	}
+
+	transposedCq := claimQueue.ToTransposed()
+
+	s.state.perRelayParent[rp] = &perRelayParentState{
+		localValidator:       localValidator,
+		statementStore:       nil, // todo use statement store (#4719)
+		session:              sessionIdx,
+		groupsPerPara:        groupsPerPara,
+		disabledValidators:   disableValidatorsSet,
+		transposedClaimQueue: transposedCq,
+		assignmentsPerGroup:  assignmentsPerGroup,
+	}
+
+	return nil
+}
+
+// Utility function to populate:
+// - per relay parent `ParaId` to `GroupIndex` mappings.
+// - per `GroupIndex` claim queue assignments
+func determineGroupAssignment(
+	numCores int,
+	groupRotationInfo *parachaintypes.GroupRotationInfo,
+	claimQueue *parachaintypes.ClaimQueue,
+) (map[parachaintypes.ParaID][]parachaintypes.GroupIndex, map[parachaintypes.GroupIndex][]parachaintypes.ParaID) {
+	// Determine the core indices occupied by each para at the current relay parent. To support
+	// on-demand parachains we also consider the core indices at next blocks.
+	ordered := claimQueue.Ordered()
+
+	schedule := make(map[parachaintypes.CoreIndex][]parachaintypes.ParaID)
+	for _, cqEntry := range ordered {
+		schedule[cqEntry.Core] = cqEntry.Paras
+	}
+
+	groupsPerPara := make(map[parachaintypes.ParaID][]parachaintypes.GroupIndex)
+	assignmentsPerGroup := make(map[parachaintypes.GroupIndex][]parachaintypes.ParaID, len(schedule))
+
+	for coreIdx, paras := range schedule {
+		groupIdx := groupRotationInfo.GroupForCore(coreIdx, uint(numCores))
+
+		assignmentsPerGroup[groupIdx] = slices.Clone(paras)
```

Comment:
We are not modifying the `paras` so i don't think we need to clone the it.
```suggestion
		assignmentsPerGroup[groupIdx] = paras
```